internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -90,73 +90,125 @@ func scenariosNamesToList(integrationTestScenarios *[]v1beta2.IntegrationTestSce
 	return &result
 }
 
-// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelines triggered by users
+// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelineruns triggered by users
 func (a *Adapter) EnsureRerunPipelineRunsExist() (controller.OperationResult, error) {
-
-	scenarioName, ok := gitops.GetIntegrationTestRunLabelValue(a.snapshot)
+	runLabelValue, ok := gitops.GetIntegrationTestRunLabelValue(a.snapshot)
 	if !ok {
-		// no test rerun triggered
 		return controller.ContinueProcessing()
 	}
-	integrationTestScenario, err := a.loader.GetScenario(a.context, a.client, scenarioName, a.application.Namespace)
 
+	integrationTestScenarios, opResult, err := a.getScenariosToRerun(runLabelValue)
+	if integrationTestScenarios == nil {
+		return opResult, err
+	}
+
+	testStatuses, err := gitops.NewSnapshotIntegrationTestStatusesFromSnapshot(a.snapshot)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	skipScenarioRerunCount, opResult, err := a.handleScenarioReruns(integrationTestScenarios, testStatuses)
+	if skipScenarioRerunCount == -1 {
```

Comment:
When does this condition appears? How can skip scenario rerun count be negative?

Reply:
> How can skip scenario rerun count be negative?

I'm returning skipScenarioRerunCount to "-1" [here](https://github.com/konflux-ci/integration-service/pull/912/files#diff-cbf7e6d58cc12a67b910b5b9108ba68d206e75cc8560ca0ab2b19e2d34415702R170).

I'm doing to pick up the situation when we have called either [StopProcessing()](https://github.com/konflux-ci/operator-toolkit/blob/ef6dcbeca69d/controller/operation.go#L87) or [RequeueWithError()](https://github.com/konflux-ci/operator-toolkit/blob/ef6dcbeca69d/controller/operation.go#L78) in any of the previous function calls. The tricky thing is `StopProcessing()` function [returns nil error](https://github.com/konflux-ci/operator-toolkit/blob/ef6dcbeca69d8d6d47ac5d58cf9183543b2deed9/controller/operation.go#L92), so we can't just use something like `if err != nil { //handle error }`.

This is why I tried to catch both of these type of calls using [if opResult.CancelRequest || err != nil](https://github.com/konflux-ci/integration-service/pull/912/files#diff-cbf7e6d58cc12a67b910b5b9108ba68d206e75cc8560ca0ab2b19e2d34415702R168) condition, and returned `-1` to help us identify this state later on.

But now that I think about it, I can replace the above condition to:
```
if opResult.CancelRequest || err != nil {
```
And it would work the same, let me know if you have any concerns about this. Thanks for making me re-think!

Reply:
Thank you sir for explanation, I'm fine with either way.